static_frame/core/index_hierarchy.py

```diff
@@ -456,211 +719,273 @@ def _from_type_blocks(cls: tp.Type[IH],
+        '''
+        def gen_blocks() -> tp.Iterator[np.ndarray]:
+            for index, indexer in zip(self._indices, self._indexers):
+                yield index.values[indexer]
+
+        return TypeBlocks.from_blocks(gen_blocks())
+
+    # --------------------------------------------------------------------------
 
-    #---------------------------------------------------------------------------
-    def __init__(self,
-            levels: tp.Union[IndexLevel, 'IndexHierarchy'],
+    def __init__(self: IH,
+            indices: tp.Union[IH, tp.List[Index]],
             *,
+            indexers: np.ndarray = EMPTY_ARRAY_INT,
             name: NameType = NAME_DEFAULT,
             blocks: tp.Optional[TypeBlocks] = None,
             own_blocks: bool = False,
-            ):
+            ) -> None:
         '''
         Initializer.
 
         Args:
-            levels: :obj:`IndexLevels` instance, or, optionally, an :obj`IndexHierarchy` to be used to construct a new :obj`IndexHierarchy`.
+            indices: list of :obj:`Index` objects
+            indexers: a 2D indexer array
+            name: name of the IndexHierarchy
+            blocks:
+            own_blocks:
         '''
+        self._recache = False
+        self._index_types = None
+        self._pending_extensions = None
 
-        self._blocks = None #type: ignore
-
-        if isinstance(levels, IndexHierarchy):
-            if not blocks is None:
-                raise ErrorInitIndex('cannot provide blocks when initializing with IndexHierarchy')
-            index_level = levels._levels
-            # if cache is updated, can get blocks
-            if not levels._recache:
-                self._blocks = levels._blocks.copy()
-            # transfer name if not given as arg
-            if name is NAME_DEFAULT:
-                name = levels.name
+        if isinstance(indices, IndexHierarchy):
+            if indexers is not EMPTY_ARRAY_INT:
+                raise ErrorInitIndex(
+                    'indexers must not be provided when copying an IndexHierarchy'
+                )
+            if blocks is not None:
+                raise ErrorInitIndex(
+                    'blocks must not be provided when copying an IndexHierarchy'
+                )
 
-        elif isinstance(levels, IndexLevel):
-            index_level = levels
-            if not blocks is None:
-                self._blocks = blocks if own_blocks else blocks.copy()
+            if indices._recache:
+                indices._update_array_cache() # pragma: no cover
+
+            self._indices = [
+                mutable_immutable_index_filter(self.STATIC, index)
+                for index in indices._indices
+            ]
+            self._indexers = indices._indexers
+            self._name = name if name is not NAME_DEFAULT else indices._name
+            self._blocks = indices._blocks
```

Comment:
Are we sure that we can take a reference here and not a shallow copy? I.e., is `self._blocks` never mutated in place?

Reply:
`TypeBlocks` is not immutable; only its underlying arrays are. What about in the case of an `IndexHierarchyGO`?

Reply:
If the array is `IndexHierarchyGO`, then `_blocks` will get updated in `_update_array_cache`, which gets called right before this move, so we can safely create a reference to `indices._blocks`